channeld/full_channel.c

```diff
@@ -1144,20 +1149,24 @@ bool channel_sending_revoke_and_ack(struct channel *channel)
 	return (change & HTLC_REMOTE_F_PENDING);
 }
 
-size_t num_channel_htlcs(const struct channel *channel)
+static inline bool any_htlc_is_dead(const struct channel *channel)
 {
 	struct htlc_map_iter it;
 	const struct htlc *htlc;
-	size_t n = 0;
 
 	for (htlc = htlc_map_first(channel->htlcs, &it);
 	     htlc;
 	     htlc = htlc_map_next(channel->htlcs, &it)) {
-		/* FIXME: Clean these out! */
-		if (!htlc_is_dead(htlc))
-			n++;
+		if (htlc_is_dead(htlc))
+			return true;
 	}
-	return n;
+	return false;
+}
+
+size_t num_channel_htlcs(const struct channel *channel)
+{
+	assert(!any_htlc_is_dead(channel));
```

Comment:
With this change should this function never return true in any context?

Even then I am not convinced we should hit an assertion for an accidental memory leak if this is violated. Print an unexpected log and continue?

Reply:
The reason for using `assert` is that the check is expensive (iterate over the entire htable) versus the actual function (return the value of a structure member). Assertions can be compiled out by defining `NDEBUG`. An unconditional check couldn't be.

The purpose of `assert` is precisely to assert a condition that is always true unless there is a bug in the program that would cause incorrect operation. That's the case here. If there are any dead HTLCs in the map, then we *cannot* safely continue, as we would then return an incorrect result from `num_channel_htlcs`.